src/v/cloud_storage/access_time_tracker.cc

```diff
@@ -54,21 +44,130 @@ void read_nested(
+        serde::write(serialize_buf, it.first);
+        serde::write(serialize_buf, it.second);
+        ++i;
+        if (i % chunk_count == 0 || i == _table.size()) {
+            for (const auto& f : serialize_buf) {
+                co_await out.write(f.get(), f.size());
+            }
+            serialize_buf.clear();
+        }
+    }
+
+    // Before dropping lock, drain any pending upserts that came in via
+    // add/remove calls while we were locking the main table.
+    for (const auto& [hash, ts] : _pending_upserts) {
+        if (ts.has_value()) {
+            _table[hash] = ts.value();
+        } else {
+            _table.erase(hash);
+        }
+    }
+    _pending_upserts.clear();
+}
+
+ss::future<> access_time_tracker::read(ss::input_stream<char>& in) {
+    _table.clear();
+    _dirty = false;
+
+    // Accumulate a serialized table_header in this buffer
+    iobuf header_buf;
+
+    // Read serde envelope header
+    auto envelope_header_tmp = co_await in.read_exactly(
+      serde::envelope_header_size);
+    header_buf.append(envelope_header_tmp.get(), envelope_header_tmp.size());
+
+    // Peek at the size of the header's serde body
+    iobuf envelope_header_buf = header_buf.copy();
+    auto peek_parser = iobuf_parser(std::move(envelope_header_buf));
+    auto header_size = serde::peek_body_size(peek_parser);
+
+    // Read the rest of the header + decode it
+    auto tmp = co_await in.read_exactly(header_size);
+    header_buf.append(tmp.get(), tmp.size());
+    auto h_parser = iobuf_parser(std::move(header_buf));
+    table_header h = serde::read_nested<table_header>(h_parser, 0);
+
+    // How many items to consume per stream read()
+    constexpr size_t chunk_count = 2048;
+
+    for (size_t i = 0; i < h.table_size; i += chunk_count) {
+        auto item_count = std::min(chunk_count, h.table_size - i);
+        auto tmp_buf = co_await in.read_exactly(item_count * table_item_size);
+        iobuf items_buf;
+        items_buf.append(std::move(tmp_buf));
+        auto parser = iobuf_parser(std::move(items_buf));
+        for (size_t j = 0; j < item_count; ++j) {
+            uint32_t hash = serde::read_nested<uint32_t>(parser, 0);
+            timestamp_t t = serde::read_nested<timestamp_t>(parser, 0);
+            _table.emplace(hash, t);
+        }
+    }
+}
+
 void access_time_tracker::add_timestamp(
   std::string_view key, std::chrono::system_clock::time_point ts) {
     uint32_t seconds = std::chrono::time_point_cast<std::chrono::seconds>(ts)
                          .time_since_epoch()
                          .count();
     uint32_t hash = xxhash_32(key.data(), key.size());
-    _table.data[hash] = seconds;
-    _dirty = true;
+
+    if (_serialization_lock.current() == 0) {
+        // Locked during serialization, defer write
+        _pending_upserts[hash] = seconds;
+    } else {
+        // Unlocked, update main table
+        _table[hash] = seconds;
+        _dirty = true;
+    }
```

Comment:
Originally, I avoided actually taking the units, because I'm not aiming for callers to mutually exclude one another, but coming back to the code after some time i agree that the act of taking the units (even if we don't use them) is more readable.